File: scripts/diagnostics/andromeda_anex_saved_audit.php
<?php
// Export only hotel identity/content evidence; never sessions, raw offer IDs or credentials.
error_reporting(0);ob_start();$pdo=null;
try {
 $root=realpath(getcwd());if(!$root||basename($root)!=='anytoour.ru')throw new RuntimeException();
 $target=$root.'/_preview/search3-anex-candidate';$private=dirname($root,2).'/.anytoour-andromeda';
 require_once $root.(is_file($root.'/data/db-v1.php')?'/data/db-v1.php':'/v2/data/db-v1.php');
 require_once $target.'/app/integrations/anex-search-mapping-registry.php';
 $pdo=v2_data_db();$pdo->setAttribute(PDO::ATTR_ERRMODE,PDO::ERRMODE_EXCEPTION);$pdo->exec('START TRANSACTION READ ONLY');
 $registry=AnyTourAnexSearchMappingRegistry::fromPdo($pdo);
 $hotels=$pdo->query('SELECT id,name,country_id,region_name,category FROM catalog_hotels WHERE country_id=4 AND is_active=1 ORDER BY id LIMIT 20001')->fetchAll(PDO::FETCH_ASSOC);
 $aliases=$pdo->query('SELECT a.hotel_id,a.alias FROM hotel_aliases a JOIN catalog_hotels h ON h.id=a.hotel_id WHERE h.country_id=4 AND h.is_active=1 ORDER BY a.hotel_id,a.alias LIMIT 50001')->fetchAll(PDO::FETCH_ASSOC);
 if(count($hotels)>20000||count($aliases)>50000)throw new RuntimeException();
 $identities=$pdo->query('SELECT supplier_namespace,external_hotel_id,local_hotel_id,decision_status,catalog_sha256,evidence_sha256,evidence_json FROM andromeda_hotel_identities ORDER BY supplier_namespace,external_hotel_id LIMIT 20001')->fetchAll(PDO::FETCH_ASSOC);
 $byId=[];foreach($identities as $i)$byId[$i['supplier_namespace'].':'.$i['external_hotel_id']]=$i;
 $files=glob($private.'/searches/*.json')?:[];if(count($files)>2000)throw new RuntimeException();
 $rows=[];$pages=0;$anexOffers=0;
 foreach($files as $file){
  if(!preg_match('/-[1-9][0-9]*\.json$/D',$file)||filesize($file)>2200000)continue;
  $state=json_decode(file_get_contents($file),true,32,JSON_THROW_ON_ERROR);
  if(!in_array($state['status']??null,['complete','partial'],true)||($state['criteria']['STATEINC']??null)!==5)continue;
  ++$pages;
  foreach($state['store']['snapshot']['offers']??[] as $o){
   if(($o['operator_ref']??null)!=='5')continue;++$anexOffers;
   $key=$o['supplier_namespace'].':'.$o['external_hotel_id'];$content=$o['hotel_content']??[];
   $image=$content['image_url']??null;$anex=null;
   if(is_string($image)&&preg_match('~^https://gateway\.samo\.ru/web/data/hotel/[0-9]+x[0-9]+/5\.([1-9][0-9]*)\.([1-9][0-9]*)\.(?:jpg|jpeg|png)$~D',$image,$m)
      &&$m[2]===$o['external_hotel_id']&&$o['supplier_namespace']==='andromeda_catalog')$anex=(int)$m[1];
   elseif($o['supplier_namespace']==='operator_5'&&ctype_digit($o['external_hotel_id']))$anex=(int)$o['external_hotel_id'];
   $identity=$byId[$key]??null;if(($identity['decision_status']??null)==='accepted')continue;
   $row=['supplier_namespace'=>$o['supplier_namespace'],'external_hotel_id'=>$o['external_hotel_id'],'hotel'=>$o['hotel'],
     'image_url'=>$image,'hotel_url'=>$content['hotel_url']??null,'region'=>$content['region']??null,
     'anex_candidate_id'=>$anex,'anex_accepted_local_id'=>$anex?$registry->resolve('anex_online',$anex,'preview'):null,
     'identity'=>$identity];
   $rowKey=$key.':'.hash('sha256',json_encode([$row['hotel'],$image]));$rows[$rowKey]=$row;
  }
 }
 $pdo->exec('ROLLBACK');
 $csp=[];foreach([$root.'/.htaccess',$target.'/.htaccess',$target.'/search-page-v2.php'] as $p)if(is_file($p))foreach(file($p) as $line)if(stripos($line,'content-security-policy')!==false||stripos($line,'img-src')!==false)$csp[]=trim($line);
 $out=['status'=>'ok','operator_filter'=>'5','supplier_calls'=>0,'database_writes'=>0,'saved_pages_read'=>$pages,'anex_offer_rows'=>$anexOffers,'rows'=>array_values($rows),'hotels'=>$hotels,'aliases'=>$aliases,'csp_source_lines'=>$csp];
}catch(Throwable $e){if($pdo instanceof PDO&&$pdo->inTransaction())$pdo->rollBack();$out=['status'=>'failed','reason'=>'saved_identity_export_failed'];}
while(ob_get_level())ob_end_clean();echo json_encode($out,JSON_UNESCAPED_UNICODE|JSON_UNESCAPED_SLASHES);

File: scripts/diagnostics/andromeda_anex_saved_audit.py
"""Correlate saved ANEX-only Andromeda hotel evidence without supplier requests."""
import hashlib,json,os
from pathlib import Path
import anex_search3_owner_decisions as owner
from andromeda_catalog_sync import names
from andromeda_hotel_candidates import save

def analyse(data):
    hotels={int(h['id']):h for h in data['hotels']}
    aliases={i:set(names(h['name'])) for i,h in hotels.items()}
    for a in data['aliases']:aliases[int(a['hotel_id'])].update(names(a['alias']))
    index={}
    for i,ns in aliases.items():
        for n in ns:index.setdefault(n,set()).add(i)
    rows=[]
    for row in data['rows']:
        ids=set()
        for n in names(row['hotel']):ids.update(index.get(n,set()))
        identity=row['identity'] or {}; existing=identity.get('decision_status')
        accepted=row['anex_accepted_local_id']; proposed=None;reason='no_unique_identity'
        if existing=='accepted':reason='already_accepted'
        elif existing=='conflict':reason='preserve_conflict'
        elif row['supplier_namespace']!='andromeda_catalog':reason='operator_namespace_needs_specific_evidence'
        elif len(ids)==1:
            candidate=next(iter(ids))
            if accepted is not None and int(accepted)!=candidate:reason='anex_name_conflict'
            else:proposed=candidate;reason='unique_egypt_current_or_former_name'+('_and_accepted_anex_id' if accepted else '')
        rows.append({**row,'name_candidates':sorted(ids),'proposed_local_id':proposed,'reason':reason,'target':hotels.get(proposed)})
    return {'operator_filter':'5','supplier_calls':0,'database_writes':0,'saved_pages_read':data['saved_pages_read'],'anex_offer_rows':data['anex_offer_rows'],'counts':{s:sum(r['reason']==s for r in rows) for s in sorted(set(r['reason'] for r in rows))},'rows':rows,'csp_source_lines':data['csp_source_lines']}

def main():
    dest=Path(os.environ['RUNNER_TEMP'])/'andromeda-anex-saved';dest.mkdir(exist_ok=True)
    save(dest/'reservation.json',{'state':'inflight','operator_filter':'5','supplier_calls':0},exclusive=True)
    source=Path(__file__).with_suffix('.php').read_text().removeprefix('<?php')
    data=owner.ssh_php(source,{},maximum_bytes=4000000)
    if data.get('status')!='ok':raise ValueError('saved identity export failed')
    save(dest/'saved-hotel-evidence.json',data,exclusive=True)
    report=analyse(data);save(dest/'anex-only-correlation.json',report,exclusive=True)
    print(json.dumps(report,ensure_ascii=False))
    print('EVIDENCE_SHA256='+hashlib.sha256((dest/'saved-hotel-evidence.json').read_bytes()).hexdigest())
if __name__=='__main__':main()
